Interactivity API and Product Button

gigitux opened the pull request:
This PR adds the Interactivity API to the Product Button to improve the UX that the user has when add a product in the cart from the Product Grid. Check the video to see the differences:

| Before | After |
|--------|--------|
|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/12ba96a4-f542-4357-bac8-3bd2e777e88a /> |<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/f8b16c63-49a7-4448-8553-3c81b8d85844 />| 
 

Thanks to Interactivity API, the Product Button:
- can show the quantity of the product in the cart.
- can update the counter syncing the cart store redux.
- doesn't load JQuery anymore.

There is a follow-up PR that adds E2E tests woocommerce/woocommerce-blocks#10036.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Ensure that you are using the Blockified Product Catalog Template. If not, enable it.
2. Open the Site Editor and add the Mini Cart Block on the header. Save.
3. Visit the Product Catalog (`/shop`).
4. Click the button to ensure the product is added to the cart.
5. Ensure that the transition Add to cart -> loading status -> 1 in the cart is smooth.
6. Click on the Mini Cart.
7. Edit the quantity of the product.
8. Ensure that the changes are reflected in the Product Button.
9. Via dev tools ensure that the button element has the class `added` when the product is in the cart, and the button element has the class `loading` while the product is added to the cart.


##### Test ProductButton with cache plugin
1. Install the plugin [WP Optimize](https://wordpress.org/plugins/wp-optimize/).
2. Ensure that your cart is empty.
3. Ensure that you have the Mini Cart block in the header.
4. Enable the cache.
5. Visit the page with the Products block.
6. Refresh the page.
7. Add a product.
8. Refresh the page.
9. Ensure that the counter is updated with a smoother animation.
10. Open the Mini Cart block and change the quantity of the product in the cart.
11. Ensure that the counter inside the button is updated with a smoother animation.
12. Disable the plugin.
13. Visit the page with the Products block.
14. Open the Mini Cart block and change the quantity of the product in the cart.
15. Ensure that the counter inside the button is updated with a smoother animation.
16. Ensure that there isn't any regression.


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Product Button block: Integrate Interactivity API to make the block dynamically.

DAreRodz committed:
Update Interactivity API JS files

DAreRodz committed:
Disable TS checks in the Interactivity API for now

DAreRodz committed:
Add new SSR files

DAreRodz committed:
Replace wp_ prefixes with wc_ ones

DAreRodz committed:
Replace wp- prefix with wc-

DAreRodz committed:
Replace guternberg_ prefix with woocommerce_

DAreRodz committed:
Remove file comments from Gutenberg

DAreRodz committed:
Rename files with `wp` prefix

DAreRodz committed:
Fix code to load Interactivity API php files

DAreRodz committed:
Remove TODO comments

DAreRodz committed:
Replace @wordpress with @woocommerce

DAreRodz committed:
Update Webpack configuration

DAreRodz committed:
Fix directive prefix

DAreRodz committed:
Remove interactivity folder from tsconfig exclude

DAreRodz committed:
Add client-side navigation meta tag code

DAreRodz committed:
Remove unneeded blocks.php file

DAreRodz committed:
Fix store tag id

DAreRodz committed:
Register Interactivity API runtime script

DAreRodz committed:
Fix Interactivity API runtime registering

DAreRodz committed:
Remove all files related to directive processing in PHP

DAreRodz committed:
Merge branch 'trunk' into update/interactivity-api

DAreRodz committed:
Move json_encode to Store's render method

gigitux committed:
Merge branch 'update/interactivity-api' of https://github.com/woocommerce/woocommerce-blocks into 6949-different-behavior-of-the-product-button-block-when-it-is-used-in-the-product-query-block-and-it-is-used-in-the-all-products-block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 6949-different-behavior-of-the-product-button-block-when-it-is-used-in-the-product-query-block-and-it-is-used-in-the-all-products-block

gigitux committed:
WIP

gigitux committed:
WIP

gigitux committed:
WIP

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
WIP

luisherranz committed:
Preserve previous context

luisherranz committed:
Ignore Minicart block on client-side navigation

luisherranz committed:
Refresh page on store updatRefresh page on store updatee

luisherranz commented:
Hey Luigi, I took a look at the synchronization of the Products blocks with David and I've pushed an update. I also took a look at your code and made some changes that I wanted to explain.

Since then, you've pushed your own changes as well, so I haven't pushed everything. I'll resolve the conflicts and push my code in a moment, but I wanted to post the explanation video before, so you can take a look:

<a href="https://www.loom.com/share/5a86600db57f40388585c3ea8817a15a">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/5a86600db57f40388585c3ea8817a15a-with-play.gif">
  </a>

https://www.loom.com/share/5a86600db57f40388585c3ea8817a15a

If any of the concepts is not clear, please let me know 🙂

luisherranz committed:
Refactor logic

luisherranz committed:
Add console error when a path is missing

luisherranz commented:
Ok, I've resolved the conflicts with your latest changes. I think everything should be working fine now, but please take a look.

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux commented:
> Hey Luigi, I took a look at the synchronization of the Products blocks with David and I've pushed an update. I also took a look at your code and made some changes that I wanted to explain.
> 
> Since then, you've pushed your own changes as well, so I haven't pushed everything. I'll resolve the conflicts and push my code in a moment, but I wanted to post the explanation video before, so you can take a look:
> 
> [ ![](https://camo.githubusercontent.com/aee9b335265376da74ac862701b11d04d43d7ba8ccaeb745572a4536cd0f2e06/68747470733a2f2f63646e2e6c6f6f6d2e636f6d2f73657373696f6e732f7468756d626e61696c732f35613836363030646235376634303338383538356333656138383137613135612d776974682d706c61792e676966) ](https://www.loom.com/share/5a86600db57f40388585c3ea8817a15a)
> https://www.loom.com/share/5a86600db57f40388585c3ea8817a15a
> 
> If any of the concepts is not clear, please let me know 🙂

Thanks, @luisherranz, for sharing the video and thoughts and helping me with the PR.


## Synchronization issue
Thanks for your incredible work. I noticed that we had a similar issue ([check this video](https://d.pr/v/odYNiG)) with the Cart block as well. Since we are updating some common utilities to update the cart (it is the only way to go), is it possible to implement a generic solution? This allows us to fix [this issue](https://github.com/woocommerce/woocommerce/issues/42352) as well.


## Missing Selectors
Yep, sorry for not using them, but I trusted the TypeScript type system, so I was convinced that it wasn't possible to use them with the current version.

![Screen Capture on 2023-06-30 at 16-55-59](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/14b0c2d6-cfe2-4251-a64f-f29e19c37171)

## Using Interactivity API only for Interactivity thing
Thank you for mentioning it. I was still working on this PR, after having the first draft, I started to write E2E tests (but after that, I had to switch to other tasks) to be sure that major refactors not introduce any regression. BTW, thanks for confirming this!

luisherranz commented:
> but I trusted the TypeScript type system

We haven't started adding support for TypeScript yet. Sorry about that. One step at a time 😄

> I noticed that we had a similar issue (check this video) with the Cart block as well. Since we are updating some common utilities to update the cart (it is the only way to go), is it possible to implement a generic solution? This allows us to fix this issue as well.

I'll take a look and I'll get back to you.

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button

gigitux committed:
Merge branch 'add/6949-interactivity-api-product-button' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

luisherranz commented:
> I was convinced that it wasn't possible to use them with the current version.

For now, `selectors` is just an arbitrary name, it doesn't have anything special. Only `state` (and `context`) are special, because they are reactive.

> I started to write E2E tests to be sure that major refactors not introduce any regression.

Yes, please. Having E2E tests would be great to make sure we don't break anything 🙂

> I noticed that we had a similar issue with the Cart block as well

Since both the Mincart and the Cart are using the same Redux store, why don't you hydrate the Minicart when the Cart is present? Wouldn't that be much simpler?

gigitux committed:
fix PHP lint error

gigitux committed:
Merge branch 'add/6949-interactivity-api-product-button' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button-store

gigitux committed:
WIP store

gigitux committed:
use store approach

gigitux committed:
Merge branch 'add/6949-interactivity-api-product-button-store' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button-store

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button

gigitux committed:
update jest configuration

gigitux committed:
Merge branch 'add/6949-interactivity-api-product-button' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
restore Mini Cart changes

gigitux committed:
move cart store subscription to interactivity package

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button

gigitux committed:
move interactivity flag

gigitux committed:
Merge branch 'add/6949-interactivity-api-product-button' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
format HTML

gigitux committed:
move addToCartText to the context

Aljullu committed:
Load product-query stylesheet when rendering the Products block

gigitux committed:
Merge branch 'fix/product-query-stylesheet' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
update sideEffects array

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button

gigitux commented on src/BlockTypes/ProductButton.php:
```diff
@@ -64,6 +97,30 @@ protected function render( $attributes, $content, $block ) {
+					'isLoading'     => false,
+					'numberOfItems' => $number_of_items_in_cart,
+					'productId'     => $product->get_id(),
+					'addToCartText' => null !== $product->add_to_cart_text() ? $product->add_to_cart_text() : __( 'Add to cart', 'woo-gutenberg-products-block' ),
```
It is necessary to put the `addToCartText` into the $context since a developer can update the text via `woocommerce_product_add_to_cart_text`.

luisherranz replied:
You mean that each product can have a different Add to Cart text, right?

gigitux replied:
Yes, potentially yes.

luisherranz replied:
Ok, thanks 🙂

gigitux commented on src/BlockTypes/ProductButton.php:
```diff
@@ -15,11 +15,20 @@ class ProductButton extends AbstractBlock {
 	 */
-	protected function register_block_type_assets() {
-		return null;
+	protected function get_block_type_script( $key = null ) {
+		$script = [
+			'handle'       => 'wc-' . $this->block_name . '-interactivity-frontend',
+			'path'         => $this->asset_api->get_block_asset_build_path( $this->block_name . '-interactivity-frontend' ),
+			'dependencies' => [],
+		];
+
+		return $key ? $script[ $key ] : $script;
```
It is necessary a double-check this. I'm not sure that it is correct!

gigitux commented:
> Since both the Mincart and the Cart are using the same Redux store, why don't you hydrate the Minicart when the Cart is present? Wouldn't that be much simpler?

In the end, we decided not to use the navigation approach but sync the Interactivity Store with the Redux Store. We followed this approach for multiple reasons:
- The navigation approach works, but it was necessary to execute an action/event on a function to update the store redux (used by ALL the entities that interact with the cart) just for the AddToCartButton.
- Slowly, as other blocks migrate to the Interactivity API, we will already have a basic infrastructure for blocks that integrate with the Cart Store.

Aljullu commented on assets/js/atomic/blocks/product-elements/button/block.json:
```diff
@@ -34,6 +34,7 @@
 			"background": false,
 			"link": true
 		},
+		"interactivity":true,
```
```suggestion
		"interactivity": true,
```

Aljullu commented on assets/js/interactivity/router.js:
```diff
@@ -1,4 +1,10 @@
+ * External dependencies
+ */
 import { hydrate, render } from 'preact';
+/**
```
```suggestion
import { hydrate, render } from 'preact';

/**
```

Aljullu commented on assets/js/interactivity/directives.js:
```diff
@@ -32,20 +32,24 @@ export default () => {
 		( {
 			directives: {
-				context: { default: context },
+				context: { default: newContext },
```
I see some TypeScript warnings in this file mentioning these values are not typed. I think it was the same before this PR, but wanted to mention anyway.

luisherranz replied:
I don't think we'll use TypeScript for the Interactivity API internals.

Do you want us to use `@ts-nocheck` or is it ok like this?

gigitux replied:
I think that it is OK to keep it as it is!

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -0,0 +1,149 @@
+				product?.quantity?.toString()
+			);
+		},
+		moreThanOneItem: ( {
```
Nit: what do you think about renaming this selector so it starts with a verb? I feel it would make it clearer what it returns. Something like: `isThereMoreThanOneItem` or `cartContainsMoreThanOneItem`?

gigitux replied:
Good point! I renamed it with [830e71f](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/830e71f1c80fdca98f94c092ef8fddbf7f16e3b8)

Aljullu commented on assets/js/interactivity/cart/cart-store.ts:
```diff
@@ -0,0 +1,28 @@
+import { updateStore } from '../store';
+/**
+ * Internal dependencies
+ */
```
```suggestion
```

Aljullu commented on src/BlockTypes/ProductButton.php:
```diff
@@ -29,6 +38,22 @@ protected function get_block_type_uses_context() {
+		add_action(
+			'wp_enqueue_scripts',
+			function() {
+				wp_dequeue_script( 'wc-add-to-cart' );
```
I think because of this line, if I render a legacy product grid block like Top Rated Products next to the Products (Beta) block, AJAX don't work in the Add to cart button and it causes a page refresh. Can you reproduce it too?

gigitux replied:
Good catch! I'm able to reproduce it. I would keep this logic. We are going to deprecate the products grid, so I prefer not to enqueue some JS on the modern blocks. Also, I think it is uncommon for old product grid blocks to be combined with a Products block. 

What do you think?

Aljullu commented on src/BlockTypes/ProductButton.php:
```diff
@@ -15,11 +15,20 @@ class ProductButton extends AbstractBlock {
+		$script = [
+			'handle'       => 'wc-' . $this->block_name . '-interactivity-frontend',
+			'path'         => $this->asset_api->get_block_asset_build_path( $this->block_name . '-interactivity-frontend' ),
+			'dependencies' => [],
```
I wonder if we should list `wc-interactivity` as a dependency of this script. And then in `assets/js/atomic/blocks/product-elements/button/block.json` import `product-button-interactivity-frontend` directly. Do you think that would work?

```JSON
	"viewScript": [
		"product-button-interactivity-frontend"
	],
```

DAreRodz replied:
In the experiment I did (the [Simple Price Filter block](https://github.com/woocommerce/woocommerce-blocks/pull/9926/commits/98c6bf4133e13477be5df6b2f11567d361964657)), I could make it work by including both the `frontend.js` script handle and the Interactivity API runtime handle in `viewScript`. That way, I didn't need to define anything in PHP to get the `frontend.js` script handle and register/enqueue it.

https://github.com/woocommerce/woocommerce-blocks/blob/98c6bf4133e13477be5df6b2f11567d361964657/assets/js/blocks/simple-price-filter/block.json#L11-L14

gigitux replied:
Updated with [de4ff24](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/de4ff2482445ce6e0506a225beb90edd4e259bb6)

Aljullu replied:
@DAreRodz I wasn't able to get your approach working. In any case, I guess we would still need to register the script, no? In which case it wouldn't make much difference code-wise.

Aljullu commented on woocommerce-gutenberg-products-block.php:
```diff
@@ -317,3 +317,4 @@ function woocommerce_blocks_interactivity_setup() {
 	}
 }
 add_action( 'plugins_loaded', 'woocommerce_blocks_interactivity_setup' );
+add_filter( 'woocommerce_blocks_enable_interactivity_api', '__return_true' );
```
Instead of doing this, could we change the default in line 312? I feel it makes it easier to understand:

```diff
function woocommerce_blocks_interactivity_setup() {
	// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
	$is_enabled = apply_filters(
		'woocommerce_blocks_enable_interactivity_api',
-		false
+		true
	);

	if ( $is_enabled ) {
		require_once __DIR__ . '/src/Interactivity/load.php';
	}
}
```

gigitux replied:
Fixed with [ef954c9](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/ef954c9dbc7aa3e6291ccdf01141c314a6e4e920)

Aljullu reviewed (commented):
Wow, outstanding work here, @gigitux! :clap: 

This is working like a charm. The only big issue I found is that if I enable a caching plugin, then the session from one user might leak into another. In this screenshot you can see I have only one product in the cart, but the Add to Cart buttons show I have more:

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/48b2ab57-c3d6-46e9-b034-61889c46d599)

I left some other comments inline below: :point_down:

DAreRodz commented on bin/webpack-entries.js:
```diff
@@ -151,6 +151,8 @@ const entries = {
 		'mini-cart-component':
 			'./assets/js/blocks/mini-cart/component-frontend.tsx',
+		'product-button-interactivity':
+			'./assets/js/atomic/blocks/product-elements/button/frontend.tsx',
```
Do you know if this is required? I guess `getBlockEntries` already returns an entry for this block, am I right? 🤔

gigitux replied:
`getBlockEntries` is configurated only for the blocks in the path: `assets/js/blocks`. If you check the implementation ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/ef954c9dbc7aa3e6291ccdf01141c314a6e4e920/bin/webpack-entries.js/#L101-L105)), we can't use the `customDir` option as well.

Since that we will remove the `atomic` block (https://github.com/woocommerce/woocommerce-blocks/issues/6038), I keep this change as it is.

DAreRodz commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -0,0 +1,149 @@
+					context.woocommerce.numberOfItems++;
+					context.woocommerce.isLoading = false;
+				} catch ( error ) {
+					context.woocommerce.numberOfItems--;
```
Have you been able to see if this works as expected? I feel `numberOfItems` is only incremented when there is no error caught. ☝️

gigitux replied:
Great catch! I updated it via [a493959](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/a493959ab9f5ce211264904d6f7ac8423e99b71f)

gigitux committed:
fix catch

gigitux committed:
rename moreThanOneItem to isThereMoreThanOneItem

gigitux committed:
improve how scripts are enqueued

gigitux committed:
update default value for the filter woocommerce_blocks_enable_interactivity_api

gigitux committed:
Update assets/js/atomic/blocks/product-elements/button/block.json

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

gigitux committed:
Update assets/js/interactivity/cart/cart-store.ts

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

gigitux committed:
fix block.json

gigitux commented:
Hi! Thanks for your review! I addressed your feedback (except for some comments regarding the `/interactivity` package. @DAreRodz should I address them as well?

The main issue left before merging this PR is the "incompatibility" with cache plugins. I created [a discussion](https://github.com/woocommerce/woocommerce-blocks/discussions/10153) to understand if it is a blocker for merging this PR or not.

luisherranz commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -0,0 +1,151 @@
+				return context.woocommerce.addToCartText;
+			}
+
+			if ( ! state.woocommerce.cart ) {
+				return state.woocommerce.inTheCartText.replace(
+					'###',
+					context.woocommerce.numberOfItems.toString()
+				);
+			}
```
Why do you need to maintain two separate implementations? Isn't the point of syncing with Redux to delegate everything to Redux, including the number of items?

gigitux replied:
The redux store isn't initiated. It is initialized only when the Mini Cart block is opened, Checkout/Cart blocks are rendered, or the action is dispatched

luisherranz replied:
Doesn't it make sense to initialize it and use it as the source of truth for all the cart state?

gigitux replied:
The thing is that currently, the store is initialized by fetching the Store API. This could cause a flickering effect (the counter of products appears after).

The best approach is that the store redux should be initialized using the data passed via PHP.

luisherranz replied:
The flickering can be solved by populating `state.woocommerce.cart` in PHP with the correct shape and information, but I think you should figure out the [caching strategy](https://github.com/woocommerce/woocommerce-blocks/discussions/10153) before attempting to do that.

luisherranz commented on assets/js/interactivity/cart/cart-store.ts:
```diff
@@ -0,0 +1,25 @@
+		select( CART_STORE_KEY ).hasFinishedResolution( 'getCartData' );
+
+	if ( isResolutionFinished ) {
+		updateStore( {
+			state: {
+				woocommerce: {
+					cart: cartData,
+				},
+			},
+		} );
```
You don't really need a special function to update the store. You can do:

```js
state.woocommerce.cart = cartData;
```

Also, this logic should not be inside the Interactivity API per se. Take into account that this package will disappear once the Interactivity API is merged with WordPress Core. 

Ideally, we should do it within a `data-wp-init` callback, which is a `useEffect` with access to state. The problem is that we should add it only once per page, and we don't have any pattern for that. We need to think about this.

luisherranz replied:
I opened a [discussion](https://github.com/WordPress/gutenberg/discussions/52514) to discuss ideas about this. Happy to hear your ideas!

gigitux replied:
> Also, this logic should not be inside the Interactivity API per se. Take into account that this package will disappear once the Interactivity API is merged with WordPress Core.

I agree, but I had some issues with this logic outside the interactivity API package. Since it is a temporary thing (since the Interactivity API will be included in Core) and since I created a dedicated folder, we could keep this approach, for now. What do you think?

luisherranz replied:
Yes, we can merge it as it is and change it later once we figure out the right pattern.

luisherranz replied:
Let's try to implement [David's idea](https://github.com/WordPress/gutenberg/discussions/52514#discussioncomment-6419346) of a special callback passed to `store`:

```js
// product-button-view.js
import { store } from "@woocommerce/interactivity";

store(
  {
    // Regular store...
  },
  {
    afterLoad: ({ state }) => {
      const reduxStore = getReduxStore();
      reduxStore.subscribe(() => {
        state.woocommerce.cart = reduxStore.getState();
      });
      dispatch(reduxStore.fetchStoreAPI()); // Fetch the store.
    },
  }
);
```

If you are concerned about the initial load performance, you can wrap the `dispatch` with a `requestIdleCallback`.

If multiple blocks need to use the same callback, you could move it to a different file. That way, it will only load once even if there are multiple blocks on the page.

```js
// add-to-cart-button-view.js
import { store } from "@woocommerce/interactivity";
import "../redux-sync.js";

store({
  // Local store needs...
});
```

```js
// add-to-cart-view.js
import { store } from "@woocommerce/interactivity";
import "../redux-sync.js";

store({
  // Local store needs...
});
```

```js
// redux-sycn.js
import { store } from "@woocommerce/interactivity";

store(
  {
    // Your global store needs...
  },
  {
    afterLoad: ({ state }) => {
      const reduxStore = getReduxStore();
      reduxStore.subscribe(() => {
        state.woocommerce.cart = reduxStore.getState();
      });
      dispatch(reduxStore.fetchStoreAPI()); // Fetch the store.
    },
  }
);
```



@gigitux: Let us know what you think! If you like the idea, we'll add it to our tasks and we'll take care of the implementation.

cc: @darerodz

gigitux replied:
I think that it makes sense. Do you think that is there a way to init the redux store only when the page is cached? 🤔

luisherranz replied:
> Do you think that is there a way to init the redux store only when the page is cached?

Yeah, that should be possible. You just have to add that information to the state:

```php
if ( isset( WC()->cart ) ) {
  // Add user info...
  wp_store([
    'state' => [
      'woocommerce' => [
        'syncStore' => false
      ]
    ]
  ]);
} else {
  // Don't add user info...
  wp_store([
    'state' => [
      'woocommerce' => [
        'syncStore' => true
      ]
    ]
  ]);
}
```

Then, use it in the `afterLoad` callback:

```js
store(
  {},
  {
    afterLoad: ({ state }) => {
      if (state.woocommerce.syncStore) {
        // Sync the store...
      }
    },
  }
);
```

DAreRodz replied:
I'd love to implement the `afterLoad` callback. 😄

Is it required to include it here? Or could I start working on it in a different PR?

luisherranz replied:
Thanks, David. I forgot to tell you yesterday 😄

Maybe different PR and then rebase/merge in this one to keep things organized?

DAreRodz replied:
I've just opened a PR with a preliminary implementation for the Interactivity Store callbacks.
- https://github.com/woocommerce/woocommerce-blocks/pull/10338

DAreRodz replied:
> If multiple blocks need to use the same callback, you could move it to a different file. That way, it will only load once even if there are multiple blocks on the page.

@luisherranz, just pointing out that we'd have to pay attention to the Webpack bundling so we are not duplicating the same callback. ☝️

luisherranz commented on assets/js/interactivity/store.js:
```diff
@@ -35,6 +35,13 @@ const getSerializedState = () => {
 const rawState = getSerializedState();
 export const rawStore = { state: deepSignal( rawState ) };
 
+export const updateStore = ( newStore ) => {
+	rawStore.state.woocommerce = {
+		...rawStore.state.woocommerce,
+		...newStore.state.woocommerce,
+	};
+};
+
```
This is not really needed.

gigitux replied:
Fixed with 5fae283eaada268eeddd06f5f5189ae2745776e7.

luisherranz commented on assets/js/interactivity/router.js:
```diff
@@ -100,8 +106,9 @@ const fetchAssets = async ( document ) => {
 
 // Fetch a new page and convert it to a static virtual DOM.
-const fetchPage = async ( url ) => {
-	const html = await window.fetch( url ).then( ( r ) => r.text() );
+const fetchPage = async ( url, options = {} ) => {
+	const html =
+		options.html || ( await window.fetch( url ).then( ( r ) => r.text() ) );
```
We can revert all the router changes in this PR because they are not needed anymore. We'll add them directly to the Interactivity API.

gigitux replied:
Can I just restore the files that are currently in the trunk?

luisherranz replied:
Yes. We'll add those changes upstream and then we'll update them here in a separate PR to keep things in context 🙂

luisherranz reviewed (commented):
Great work with the new implementation 🙂

A couple of comments:

- I don't like the idea of maintaining two separate implementations. For example, `context.woocommerce.numberOfItems` should not exist because that information is on Redux. Is that not possible?
- As we are not going to use the navigation approach anymore, we can revert all the changes we made to the router.

gigitux committed:
remove updateStore function

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
restore interactivity api changes

gigitux committed:
import cart store

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -0,0 +1,151 @@
+				} catch ( error ) {
+					// we don't care about errors blocking execution, but will console.error for troubleshooting.
+					// eslint-disable-next-line no-console
+					console.error( error );
```
I noticed there is a different behavior when trying to add a product that can only be added once and it's already in the cart:

* In `trunk`, I get redirected to the product page and see a notice:
![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/dd0f88f2-752a-4306-a381-2e9ee537417d)
* In this branch, the button is disabled and stays in a loading state indefinitely:
![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/5438e5d7-1d61-4e00-afea-6e2cae2e6d94)

Can you reproduce as well? I don't know what's the best approach here, but we might want to improve how errors are communicated to the shopper.

gigitux replied:
Nice catch! Given that we are using the Store API, we can't rely on the PHP to generate the notice and show it on the product page. What do you think about the approach that I introduced with [efbfe3b](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/efbfe3bd274d4989955e09c89225fd0e960f77fd)

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/6e79f635-9354-4575-8a29-49657d50eaf8

Aljullu replied:
Looks perfect! The only issue I found is that the notice doesn't show up if the Store Notices block is not on the page. I think that would be an issue when adding the Products block to a post or page, for example, as the Store Notices block can't be added in that context.

gigitux replied:
Good point... if the notice block isn't added to the page, we could create a div on top of the block and inject the notice. What do you think?

Aljullu replied:
Yup, makes sense to me. We also have the [`StoreNoticesContainer` component](https://github.com/woocommerce/woocommerce-blocks/blob/55cc42e761a5c509bb2c5fae95cd241c497c213d/packages/checkout/components/store-notices-container/index.tsx#L29), but I don't know if it would be too much for this use-case.

gigitux replied:
I improved the logic with 72ff24763471494c97837bdb45caa8e10b2854d9

luisherranz replied:
Is this something that the upcoming auto-insert mechanism could handle?

- https://github.com/WordPress/gutenberg/pull/51449

If not, could it use the Interactivity API instead of a React component?

Ideally, all interactive blocks should end up using the Interactivity API so we can unblock the full-page client-side navigations.

gigitux replied:
Currently, the block that handles the notices works only with the server-side generated notice, so the auto-insert mechanism can't help for this use-case.

> If not, could it use the Interactivity API instead of a React component?

Can you expand this idea?

luisherranz replied:
Oh, I thought it was a client-side component. Sorry.

Aljullu reviewed (commented):
Thanks for all the updates, @gigitux! I was taking another look at this PR and found a couple of other edge cases which might be worth looking into. :point_down: 

In classic themes, adding one product to the cart adds two items:

[Enregistrament de pantalla des de 2023-07-13 11-51-47.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/ed5eb8a3-e9ff-4c3b-a41b-1070db608797)

I could reproduce in Storefront and Twenty Fifteen, but not in Twenty Twenty-Three. :confused:  Any idea what might be going on?

gigitux committed:
show notice when there is an error

gigitux committed:
add logic to dequeue script on classic themes and block themes

luisherranz commented:
> [9bea193](https://github.com/woocommerce/woocommerce-blocks/commit/9bea193f9b26c8e1bcfd046665aff011330803fb) Enable Interactivity API only on block theme

May I know what is preventing us from using the Interactivity API in classic themes? 🙂

gigitux commented:
> > [9bea193](https://github.com/woocommerce/woocommerce-blocks/commit/9bea193f9b26c8e1bcfd046665aff011330803fb) Enable Interactivity API only on block theme
> 
> May I know what is preventing us from using the Interactivity API in classic themes? 🙂

I'm working on [this issue](https://github.com/woocommerce/woocommerce-blocks/pull/10006#pullrequestreview-1528084584), I just realized that 9bea193f9b26c8e1bcfd046665aff011330803fb isn't the best one :D. I'm trying to find another solution!

luisherranz commented:
Ok. Thanks, Luigi. If you find any blockers running the Interactivity API in classic themes please let us know 🙂

gigitux commented:
> Thanks for all the updates, @gigitux! I was taking another look at this PR and found a couple of other edge cases which might be worth looking into. 👇
> 
> In classic themes, adding one product to the cart adds two items:
> 
>  Enregistrament.de.pantalla.des.de.2023-07-13.11-51-47.webm 
> I could reproduce in Storefront and Twenty Fifteen, but not in Twenty Twenty-Three. 😕 Any idea what might be going on?

Yes! On Classic Theme, the `add-to-cart` scripts were enqueued. I updated the logic to dequeue the scripts. 07c86ab8ea71c3656f2287154f14ddf694ff1486

gigitux committed:
imrpove logic about notice

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -0,0 +1,195 @@
+				state: State;
+				ref: HTMLElement;
+			} ) => {
+				if ( ! ref.classList.contains( 'ajax_add_to_cart' ) ) {
```
I we do that, we would need to define a difference class name that we can use to know if the button needs to work with the interactivity API.

gigitux replied:
https://github.com/woocommerce/woocommerce-blocks/pull/10006?notification_referrer_id=NT_kwDOAEQaRrI2ODkyNzI4NDY0OjQ0NjMxNzQ#discussion_r1258037507

As I said above, I think it is uncommon for old product grid blocks to be combined with a Products block. There are extensions/themes that could have specific behavior when `ajax_add_to_cart` is added. I think for compatibility reasons, it makes sense to keep the same class name.

Aljullu replied:
> There are extensions/themes that could have specific behavior when ajax_add_to_cart is added. I think for compatibility reasons, it makes sense to keep the same class name.

Oh, right, that makes sense.

> As I said above, I think it is uncommon for old product grid blocks to be combined with a Products block.

Ups, sorry if I missed this discussion. I have to say I'm not 100% sold on this part of the discussion. I think there are several use cases where a store owner might want to use the Products block alongside the Classic Template block, other WC product grid blocks or shortcodes. Ie: having the Product Catalog template with the Classic Template block but adding a banner on top with "Products on Sale" powered by the Products block. What do you think?

I don't want to block the whole PR because of this, but one thing we could explore is having a specific class for buttons using the Interactivity API (ie: `.wc-block-interactivity-api`) in addition to `.ajax_add_to_cart`. Then, updating this line in WC core:

https://github.com/woocommerce/woocommerce/blob/12e5edf03231abe6a3f98de002616a91cf369eca/plugins/woocommerce/client/legacy/js/frontend/add-to-cart.js#L63

To something like:

```JS
if ( $thisbutton.is( '.ajax_add_to_cart:not(.wc-block-interactivity-api)' ) ) {
```

(I didn't test it, so it might be more complex than what I think, but wanted to share this idea in case you think it makes sense)

gigitux replied:
After an internal discussion p1691054560419849-slack-C02FL3X7KR6, we decided to **not** work on this issue.
We will mention this caveat in the release notes so we’ve communicated it as a known issue and how they can fix it.

Aljullu reviewed (commented):
Thanks for all the updates, @gigitux! I noticed that if the Products block is rendered on the same page alongside the Classic Template block, the AJAX Add to Cart behavior doesn't work in the Classic Template block. Can you reproduce as well? I wonder if removing the `ajax_add_to_cart` class from the button when using the Interactivity API would fix this.

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

Aljullu reviewed (approved):
LGTM, outstanding work with that Luigi, Luís and David!

I left one last comment on an ongoing discussion, but besides that, everything looks good on my end.

imanish003 commented:
I am going to bump this PR for the next release as it doesn't have a blocker label.

DAreRodz committed:
Interactivity API: add `afterLoad` callbacks to `store()` function (#10338)

* show notice when there is an error

* Add initial implementation for store callbacks

* Run `afterLoad` callbacks after `init`

* Move cart state subscription to Product button

* Remove cart-store from Interactivity API internals

* Change callbacks with options and save only afterLoad callbacks

* ProductButton: Add animation (#10351)

* implement animation

* improve logic

* refactor logic

* refactor code

* address feedback about code style

* add support for woocommerce_add_to_cart_quantity

* Fix animation flickering

* Introduce wp-effect, reduce the amount of numberOfItem variables to 2 and consolidate animation status

* add support for added class

* Remove unnecessary selector

* Don't fetch cart if it was already fetched

* remove added class

---------

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

---------

Co-authored-by: Luigi <gigitux@gmail.com>
Co-authored-by: Luis Herranz <luisherranz@gmail.com>

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
update deepsignal

luisherranz commented on src/BlockTypes/ProductButton.php:
```diff
@@ -50,6 +92,7 @@ protected function render( $attributes, $content, $block ) {
 			$text_align_styles_and_classes = StyleAttributesUtils::get_text_align_class_and_style( $attributes );
 			$classname                     = $attributes['className'] ?? '';
 			$custom_width_classes          = isset( $attributes['width'] ) ? 'has-custom-width wp-block-button__width-' . $attributes['width'] : '';
+			$is_added_class                = $more_than_one_item ? 'added' : '';
```
Is this still neded?

gigitux replied:
No, I removed it with [5432309](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/5432309569b62cc84433274ed6b992b9ca4ec8fe)

luisherranz reviewed (approved):
I think this is great, Luigi. Great work 🙂👏

I'd like to take a look at how we can migrate the `NoticeBanner` to the Interactivity API, but we can do so in a new PR.

luisherranz commented on src/BlockTypes/ProductButton.php:
```diff
@@ -29,6 +38,32 @@ protected function get_block_type_uses_context() {
+	 */
+	protected function enqueue_assets( array $attributes ) {
+		parent::enqueue_assets( $attributes );
+		if ( wc_current_theme_is_fse_theme() ) {
```
One question: what's the reasoning behind enqueuing this script only on block themes? What prevents us from using it in classic themes?

cc: @gigitux

luisherranz replied:
Oh, you're dequeuing other scripts, not enqueuing this one, are you?

gigitux replied:
> Oh, you're dequeuing other scripts, not enqueuing this one, are you?

Yes, exactly! @luisherranz

luisherranz replied:
Ok, thanks, Luigi!

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/6949-interactivity-api-product-button

gigitux committed:
remove added class

gigitux committed:
update deepsignal

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button

gigitux committed:
Interactivity API and Product Button: Add E2E tests (#10036)

* Add FrontendUtils class

* fix conflicts

* use locator

* restore click usage

* Product Button: Add E2E test

* fix util

* fix E2E tests

* remove comment

* Add E2E test to ensure that woocommerce_product_add_to_cart_text works

* update sideEffects array

* add zip and unzip as package

* fix wp-env configuration

* fix E2E test

* add report

* try now

* try now

* try now

* fix E2E test

* E2E: Add documentation for testing actions and filters. Fixes #10135 (#10206)

* update description

* fix label

* rename files

* make requestUtils private

* remove page.goto

* use toHaveCount

* use productsToDisplay variable

* fix E2E tests

* rename class utils

---------

Co-authored-by: Daniel Dudzic <daniel.dudzic@automattic.com>

gigitux committed:
Merge branch 'trunk' into add/6949-interactivity-api-product-button
